### DISCUSSION TITLE
LILYGO T-ETH-Lite-ESP32S3 and ST7796S Display - Whitescreen problem

### DISCUSSION THREAD

**Discussion by Queequack** [Bodmer/TFT_eSPI#3509] [Compatible displays and setup files]
### What I wanted to do:
The prerequisite is the use of a [T-ETH-Lite with ESP32-S3](https://www.lilygo.cc/products/t-eth-lite?variant=43120880779445). I am looking for a 3.5 inch display that runs as smoothly as possible. [This discussion](https://www.lilygo.cc/products/t-eth-lite?variant=43120880779445) led me to the **ST7796** display. I wanted to use it via **4 line SPI** at 80 MHz. I also wanted to activate **DMA**, as described [here](https://www.lilygo.cc/products/t-eth-lite?variant=43120880779445) for the ESP32-S3.

I connected the 40-pin FPC of the ordered [ST7796 LCD with Capacitive Touch](https://de.aliexpress.com/item/32814953314.html?spm=a2g0o.order_list.order_list_main.25.4be25c5fyN75kB&gatewayAdapt=glo2deu) according to the [Datasheet](https://github.com/user-attachments/files/17380726/Datasheet.pdf) and the pins, especially for SPI communication, according to the [ESP32-S3-Box Setup](https://github.com/Bodmer/TFT_eSPI/blob/master/User_Setups/Setup251_ESP32_S3_Box.h) (see screenshots below). LilyGo provides the ESP32-S3-Box in the [Platformio.ini for the T-ETH-Lite-ESP32S3](https://github.com/Xinyuan-LilyGO/LilyGO-T-ETH-Series/blob/master/platformio.ini). I don't use an SD card or Ethernet. 

### Problem: 
To ensure that my code (see below) works, I have connected an RGB LED. This flashes as intended in the code. Otherwise, the code only intends to color the display red. To see a change, the display color should change with every LED action. **The display simply remains white** (Backlight switched on).

I use VSCode with PlatformIO and the Arduino Framework.

### Code:

**User_setup.h:**
```
#define USER_SETUP_INFO "User_Setup"

#define ST7796_DRIVER

#define TFT_MISO  0
#define TFT_MOSI  6
#define TFT_SCLK  7
#define TFT_CS    5
#define TFT_DC    4
#define TFT_RST  15

#define LOAD_GLCD   // Font 1. Original Adafruit 8 pixel font needs ~1820 bytes in FLASH
#define LOAD_FONT2  // Font 2. Small 16 pixel high font, needs ~3534 bytes in FLASH, 96 characters
#define LOAD_FONT4  // Font 4. Medium 26 pixel high font, needs ~5848 bytes in FLASH, 96 characters
#define LOAD_FONT6  // Font 6. Large 48 pixel font, needs ~2666 bytes in FLASH, only characters 1234567890:-.apm
#define LOAD_FONT7  // Font 7. 7 segment 48 pixel font, needs ~2438 bytes in FLASH, only characters 1234567890:-.
#define LOAD_FONT8  // Font 8. Large 75 pixel font needs ~3256 bytes in FLASH, only characters 1234567890:-.
//#define LOAD_FONT8N // Font 8. Alternative to Font 8 above, slightly narrower, so 3 digits fit a 160 pixel TFT
#define LOAD_GFXFF  // FreeFonts. Include access to the 48 Adafruit_GFX free fonts FF1 to FF48 and custom fonts

#define SMOOTH_FONT

#define SPI_FREQUENCY  27000000
```

**main.cpp:**

```
#include <Arduino.h>
#include <TFT_eSPI.h>
#include <FastLED.h>

// ################## Display ##################
TFT_eSPI tft = TFT_eSPI();
// #############################################

// ############### Flashing LED ###############
CRGB rgbLed[1];
bool ledState = true;
unsigned long lastTime = 0;
unsigned long blinkTime = 500;
// #############################################

bool timerExpired(unsigned long, unsigned long);

void setup()
{
  // ################## Display ##################
  tft.init();
  tft.setRotation(1);
  tft.fillScreen(TFT_RED);
  //tft.fillScreen(TFT_BLACK);
  //tft.setCursor(0,0,4);
  //tft.setTextColor(TFT_WHITE);
  //tft.println ("Hello World!");
  // #############################################

  // ############### Flashing LED ###############
  FastLED.addLeds<SK6812, 38, GRB>(rgbLed, 1);
  FastLED.setBrightness(50);
  rgbLed[0] = CRGB::Black; FastLED.show();
  ledState = false;
  // #############################################
}

void loop() 
{
  // ############### Flashing LED ###############
  if (timerExpired(lastTime, blinkTime))
  {
    lastTime = millis();
    if(ledState) { rgbLed[0] = CRGB::Black; FastLED.show(); ledState = false; tft.fillScreen(TFT_RED);   }
    else         { rgbLed[0] = CRGB::Violet; FastLED.show(); ledState = true; tft.fillScreen(TFT_BLACK); }
  }
  // #############################################
}

/*********************************************************************************************\
 * Timer */
bool timerExpired(unsigned long last, unsigned long del)
{
  if (millis() - last >= del) return true;
  return false;
}
```

**platformio.ini:**

```
[platformio]
default_envs = T-ETH-Lite-ESP32S3

[env]
platform = espressif32@6.4.0
framework = arduino
upload_speed =  921600
monitor_speed = 115200
;monitor_filters =
	;default
	;esp32_exception_decoder

;build_flags =
;	-DCORE_DEBUG_LEVEL=1

;Use OTA update firmware
; upload_protocol = espota
; upload_port = 192.168.36.107
    
; ESP32-S3-WROOM-1 + W5500 FLASH:16MB PSRAM:8MB OPI
[env:T-ETH-Lite-ESP32S3]
board=esp32s3box
build_flags = 
	${env.build_flags}
	-DBOARD_HAS_PSRAM
    -DLILYGO_TETH_POE
	-DLILYGO_T_ETH_LITE_ESP32S3
    -UARDUINO_USB_CDC_ON_BOOT
	-D USER_SETUP_LOADED=1
    -include $PROJECT_LIBDEPS_DIR/$PIOENV/TFT_eSPI/User_Setup.h

lib_ignore = 
    Ethernet

lib_deps =
    fastled/FastLED @ ^3.6.0
	;bodmer/TFT_eSPI @ ^2.5.43 ; Display Treiber
	TFT_eSPI
```

### Pin assignment

![Pin](https://github.com/user-attachments/assets/827e2158-0286-48eb-9331-ab76a5d82d48)

![T-ETH-Lite](https://github.com/user-attachments/assets/440d663e-279a-4619-92ba-5cf99a734f83)

Regards René

**Comment by Basitadam** [comment]
Hi,
Which lines are uncommented in your "User_Setup_Select.h" file?

**Reply by Queequack** [reply]
Hi, 

`#include <User_Setup.h>`

User_Setup.h is included via platformio.ini:
```
build_flags = 
     ;...
     -include $PROJECT_LIBDEPS_DIR/$PIOENV/TFT_eSPI/User_Setup.h
```
I've also tested Setup251_ESP32_S3_Box.h with ST7796_DRIVER (instead of ILI9341_DRIVER).

Regards

**Comment by Basitadam** [comment]
What are you getting from the serial port?

1- Disconnect MISO line for now and modify to this "#define TFT_MISO -1" in the User_Setup.h
2- Add this to the User_Setup.h :  #define USE_HSPI_PORT

Now what are you getting from the serial port?

BR,
Tim

**Comment by Queequack** [comment]
Hi Tim, thanks for your help!

I have found the solution. I was convinced that the RESET pin can be selected at will. Since IO48 is not led out via the pin header on the T-ETH-Lite, I selected IO15. Both work:
- I have set TFT_RST -1 and connected TFT_RST to RST of the ESP32-S3. 
- Then I used TFT_RST 48 and held the jumper cable to the pin by hand.

I've disconnected MISO (defined -1), I won't need it anyhow. And #define USE_HSPI_PORT had no effect (even with the working solution with the RST pin).

I think the thread can be closed. May it help other lost souls.

Regards René

PS: If you have a tip on how I can run TFT_eSPI (later also with LVGL) and the rest of my program on the two seperate CPUs of the ESP32-S3, let me know.

**Reply by Basitadam** [reply]
I did a small test on dual core ( it is not recommended if you are using bluetooth and/or wifi):
`
#include <TFT_eSPI.h>                     // Library for drawing on the display
TFT_eSPI    tft     = TFT_eSPI();         // Invoke custom library
TFT_eSprite Needle  = TFT_eSprite(&tft);  // Sprite object for needle
TFT_eSprite Bgnd    = TFT_eSprite(&tft);  // Sprite object

// uncomment only one of the header files to choose between two color themes - VIOLET or RED
#include "Tekli.h"                        // images for boost gauge graphics
//#include "boost_gauge_images_VIOLET.h"  // all 61 images for boost gauge graphics

#define BasilmayacakRenk TFT_BLACK

// set the mode
int mode = 1; // 0 = ping pong animation, 1 = get value from boost pressure sensor

int frame = 0;      // current frame
int frame_inc = 1;  // frame increment

uint16_t iks,ye,sayici;
uint8_t Core0Greet=1;

float psi_older[20]; // store a few older readings for smoother readout

TaskHandle_t Task1;
TaskHandle_t Task2;


void setup() 
{
  Serial.begin(115200); 

  tft.begin();                // initialize the display
  //tft.setRotation(1);	      // rotation of the display set to landscape
  tft.setRotation(1);         // in this rotation, the USB port is on the "bottom" of the screen
  tft.setSwapBytes(true);     // Swap the colour byte order when rendering
  tft.fillScreen(TFT_BLACK);  // fill display with the black color

  CrateBackGround();
  CreateNeedle();
  tft.setPivot(120, 120);                             // Adjust middle for the TFT pivot point

  Bgnd.pushImage(0, 0, 240, 240, epd_bitmap_Taban);   // draw the fullscreen boost gauge image 
  Needle.pushRotated(&Bgnd,iks,TFT_BLACK);            // Plot rotated Sprite, black being transparent from original image
  Bgnd.pushSprite(0,0);


  pinMode(4, INPUT);                                  // GPIO4 as analog input 
  analogReadResolution(12);                           //set the resolution of analog input to 12 bits (0-4096)


 
  xTaskCreatePinnedToCore(
                    Core1code,   /* Task function. */
                    "Core1",     /* name of task. */
                    20000,       /* Stack size of task */
                    NULL,        /* parameter of the task */
                    2,           /* priority of the task, high the number = more priority */
                    &Task1,      /* Task handle to keep track of created task */
                    1);          /* pin task to core 1 */                  
  delay(500); 

  
  xTaskCreatePinnedToCore(
                    Core0code,   /* Task function. */
                    "Core0",     /* name of task. */
                    10000,       /* Stack size of task */
                    NULL,        /* parameter of the task */
                    1,           /* priority of the task */
                    &Task2,      /* Task handle to keep track of created task */
                    0);          /* pin task to core 0 */
    delay(500); 
}



void Core1code( void * pvParameters )
{
  Serial.print("Graphics is running on core ");
  Serial.println(xPortGetCoreID());

  while(Core0Greet);                      // wait for the core0 do its greeting
  

 while(1)
 {
  sayici++;
 if (mode == 0) 
   { // 0 = ping pong animation

    // play the animation - ping-pong - going from min to max and than back from max to min
    frame = frame + frame_inc;
    if ((frame > 60) && (frame_inc > 0)) {
      frame = 60;
      frame_inc = -1;
    } else if ((frame < 0) && (frame_inc < 0)) {
      frame = 0;
      frame_inc = 1;
    } 
   } 
   else if (mode == 1) 
   { // 1 = get value from boost pressure sensor

    int pot_value = analogRead(4);                                    // read analog in value from the boost pressure sensor 0-4096 

    float calculated_bar_pressure = (pot_value - 587.0) / 631.0;      // calculate BAR pressure based on the voltage, this calculation was done based on measuring analog gauge
    float calculated_psi_pressure = calculated_bar_pressure * 14.504; // calculate PSI pressure from BAR pressure

    // store a few older measurements
    for (int i = 0; i < 9; i++) {
      psi_older[i] = psi_older[i+1];
    }
    psi_older[9] = calculated_psi_pressure;

    float psi_combined = 0;
    for (int i = 0; i < 10; i++) {
      psi_combined = psi_combined + psi_older[i];
    }
    psi_combined = psi_combined / 10; // final pressure value is the average of last 10 readings

    // set the image to be displayed based on the PSI pressure,
    // we have 60 images, and the PSI pressure should be in the range -15 .. 45, i.e., also 60 steps
    int image_pressure = round(psi_combined) + 15;
    image_pressure = constrain(image_pressure, 0, 60); // restrict the images to 0-60 in case the PSI pressure is outside the range of -15..45

      iks=185+(image_pressure*3);

      Bgnd.pushImage(0, 0, 240, 240, epd_bitmap_Taban);       // draw the fullscreen boost gauge image 
      Needle.pushRotated(&Bgnd,iks, TFT_BLACK);               // Plot rotated Sprite, black being transparent from original image
      Bgnd.pushSprite(0,0);

      delay(20);      // 24 fps delay. Let core0 do its tasks also
    }
 }
}



void Core0code( void * pvParameters )
{
  Serial.print("Serial   is running on core ");
  Serial.println(xPortGetCoreID());
  Serial.println("");
  
  delay(10);    // let core0 do its greeting
  Core0Greet=0;

 while(1)
  {
    Serial.println(sayici);
    delay(700);
  }
}


void loop() 
{
 NOP();
}





void CrateBackGround(void)
{
  Bgnd.setSwapBytes(true);
  Bgnd.setColorDepth(16);
  Bgnd.createSprite(240,240);                      // create the needle Sprite
  Bgnd.fillSprite(TFT_BLACK);                      // in init tft background already filled in black
  Bgnd.pushImage(0, 0, 240, 240, epd_bitmap_Taban); // draw the fullscreen boost gauge image 
  Bgnd.pushSprite(0,0);
}

void CreateNeedle(void)
{
  Needle.setSwapBytes(true);
  Needle.setColorDepth(16);
  Needle.createSprite(135,68);                      // create the needle Sprite
  Needle.fillSprite(TFT_BLACK);                         // Color that will be NOT displayed from sprite
  uint16_t piv_x = 41;                              // x pivot of Sprite
  uint16_t piv_y = 34;                              // y pivot of Sprite (middle)
  Needle.setPivot(piv_x, piv_y);                    // Set pivot point in this Sprite
  Needle.pushImage(0,0,135,68,epd_bitmap_gauge_needle);
}

`

[ESP32_S3_Zero_Sprite_RotatedNeedleDualCore.txt](https://github.com/user-attachments/files/17393652/ESP32_S3_Zero_Sprite_RotatedNeedleDualCore.txt)


Tekli.h is an image file. A 240x240 pixel bmp image transferred to RGB565

**Reply by Queequack** [reply]
Thank you very much, I will try it out soon.
